crates/metaslang/cst/src/query/parser.rs

```diff
@@ -289,10 +332,61 @@ fn text_token(i: &str) -> IResult<&str, String, VerboseError<&str>> {
     .parse(i)
 }
 
-fn ellipsis_token(i: &str) -> IResult<&str, &str, VerboseError<&str>> {
-    terminated(tag("..."), multispace0).parse(i)
-}
-
 fn token<'input>(c: char) -> impl Parser<&'input str, char, VerboseError<&'input str>> {
     terminated(char(c), multispace0)
 }
+
+fn adjacency_operator<T: KindTypes>(i: &str) -> IResult<&str, ASTNode<T>, VerboseError<&str>> {
+    // An adjacency operator is a single '.' character, and cannot be followed
+    // by another adjacency operator
+    pair(token('.'), cut(peek(none_of(". \t\r\n"))))
+        .map(|_| ASTNode::Adjacency)
+        .parse(i)
+}
+
+fn recognize_as_failure<I: Clone, O1, O2, E: nom::error::ParseError<I>, F>(
+    mut parser: F,
+) -> impl FnMut(I) -> IResult<I, O2, E>
+where
+    F: nom::Parser<I, O1, E>,
+{
+    use nom::error::{make_error, ErrorKind};
+    use nom::Err::Failure;
+    move |input: I| {
+        let i = input.clone();
+        match parser.parse(i) {
+            Ok((_, _)) => Err(Failure(make_error(input, ErrorKind::Fail))),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+fn ellipsis_token<O>(i: &str) -> IResult<&str, O, VerboseError<&str>> {
+    use nom::bytes::complete::tag;
+    use nom::error::context;
+    context(
+        "deprecated ellipsis operator",
```

Comment:
Suggestion:

```suggestion
        "The ellipsis `...` operator is deprecated. For more information, check the [Tree Query Language](link) guide on the newly introduced adjacency operator.",
```

Reply:
I'll make this change, but this is probably not the best place to put this text. Since this is the parser context, it will read weird in the final error message. Something like:
```
Parse error:
Fail at: ...]
in section 'The ellipsis `...` operator is deprecated, and replaced with a new adjacency `.` operator. For more information, check the [Tree Query Language](https://nomicfoundation.github.io/slang/user-guide/tree-query-language/) guide.', at: ...]
```
Unfortunately, the default error types from `nom` don't allow much further customization. AFAIU, we would need to use our own custom error type to improve on this.

Reply:
I see. Thanks for the note. Maybe we can just use a normal link then, instead of using markdown:

> The ellipsis `...` operator is deprecated, and replaced with a new adjacency `.` operator. For more information, check the Tree Query Language guide: <https://nomicfoundation.github.io/slang/user-guide/tree-query-language/>

Reply:
:heavy_check_mark: